docs/advanced/UsageWithTypescript.md

```diff
@@ -0,0 +1,280 @@
+    type: ChatActions.SendMessage,
+    payload: newMessage
+  }
+}
+```
+
+System Action Constants & Shape:
+
+```ts
+// src/store/system/types.ts
+
+export enum SystemActions {
+  UpdateSession = 'UPDATE_SESSION'
+}
+
+interface UpdateSessionAction {
+  type: SystemActions.UpdateSession
+  payload: SystemState
+}
+
+export type SystemActionTypes = UpdateSessionAction
+```
+
+With these types we can now also type check system's action creators:
+
+```ts
+// src/store/system/actions.ts
+
+import { SystemActions, SystemState } from './types'
+
+export function updateSession(newSession: SystemState) {
+  return {
+    type: SystemActions.UpdateSession,
+    payload: newSession
+  }
+}
+```
+
+## Type Checking Reducers
+
+Reducers are just pure functions that take the previous state, an action and then return the next state. In this example, we explicitly declare the type of actions this reducer will receive along with what it should return (the appropriate slice of state). With these additions TypeScript will give rich intellisense on the properties of our actions and state. In addition, we will also get errors when a certain case does not return the `ChatState`.
+
+Type checked chat reducer:
+
+```ts
+// src/store/chat/reducers.ts
+
+import { ChatState, ChatActions, ChatActionTypes } from './types'
+
+const initialState: ChatState = {
+  messages: []
+}
+
+export function chatReducer(
+  state = initialState,
+  action: ChatActionTypes
+): ChatState {
+  switch (action.type) {
+    case ChatActions.SendMessage:
+      return {
+        messages: [...state.messages, action.payload]
+      }
+    default:
+      return state
+  }
+}
+```
+
+Type checked system reducer:
+
+```ts
+// src/store/system/reducers.ts
+
+import { SystemActions, SystemState, SystemActionTypes } from './types'
+
+const initialState: SystemState = {
+  loggedIn: false,
+  session: '',
+  userName: ''
+}
+
+export function systemReducer(
+  state = initialState,
+  action: SystemActionTypes
+): SystemState {
+  switch (action.type) {
+    case SystemActions.UpdateSession: {
+      return {
+        ...state,
+        ...action.payload
+      }
+    }
+    default:
+      return state
+  }
+}
+```
+
+We now need to generate the root reducer function, which is normally done using `combineReducers`. Note that we do not have to explicitly declare a new interface for AppState. We can use `ReturnType` to infer state shape from the `rootReducer`.
+
+```ts
+// src/store/index.ts
+
+import { systemReducer } from './system/reducers'
+import { chatReducer } from './chat/reducers'
+
+const rootReducer = combineReducers({
+  system: systemReducer,
+  chat: chatReducer
+})
+
+export type AppState = ReturnType<typeof rootReducer>
+```
+
+## Usage with React-Redux
+
+While react-redux is a separate library from redux itself, it is commonly used with react. For this reason, we will go through how react-redux works with TypeScript using the same example used previously in this section.
```

Comment:
Capitalize React-Redux everywhere.

Reply:
Actually, it's "React Redux". No dash, that's just the package name.